### PR TITLE
[widget audit] TextInput, PasswordInput

### LINKED ISSUE #1290: On Android, the application complains about `AttributeError: 'TextInput' object has no attribute '_validators'`
**Describe the bug**

I use a `field = toga.TextInput()` and then use `field.validate()` inside the `on_change` callback method, in order to react accordingly. The application seems to work without problems, however when I check the logs via `adb logcat -s MainActivity:* stdio:* Python:*` then I see the following errors:

```
06-25 00:41:56.036 28676 28676 D Python  : Native invocation 511594544624 :: beforeTextChanged
06-25 00:41:56.036 28676 28676 D Python  : Native invocation done.
06-25 00:41:56.036 28676 28676 D Python  : Native invocation 511594544624 :: onTextChanged
06-25 00:41:56.037 28676 28676 D Python  : Native invocation done.
06-25 00:41:56.037 28676 28676 D Python  : Native invocation 511594544624 :: afterTextChanged
06-25 00:41:56.037 28676 28676 E Python  : Traceback (most recent call last):
06-25 00:41:56.037 28676 28676 E Python  :   File "/data/user/0/what.ever.acc_companion/files/python/user_code/app_packages/rubicon/java/api.py", line 45, in dispatch
06-25 00:41:56.037 28676 28676 E Python  :     val = getattr(pyinstance, method)(*args)
06-25 00:41:56.037 28676 28676 E Python  :   File "/data/user/0/what.ever.acc_companion/files/python/user_code/app_packages/toga_android/widgets/textinput.py", line 24, in afterTextChanged
06-25 00:41:56.037 28676 28676 E Python  :     self.interface.on_change(widget=self.interface)
06-25 00:41:56.037 28676 28676 E Python  :   File "/data/user/0/what.ever.acc_companion/files/python/user_code/app_packages/toga/handlers.py", line 55, in _handler
06-25 00:41:56.037 28676 28676 E Python  :     result = handler(interface, *args, **kwargs)
06-25 00:41:56.037 28676 28676 E Python  :   File "/data/user/0/what.ever.acc_companion/files/python/user_code/app/acc_companion/utils.py", line 34, in wrapper
06-25 00:41:56.037 28676 28676 E Python  :     return func(*args, **kwargs)
06-25 00:41:56.038 28676 28676 E Python  :   File "/data/user/0/what.ever.acc_companion/files/python/user_code/app/acc_companion/app.py", line 109, in energy_changed
06-25 00:41:56.038 28676 28676 E Python  :     if field.validate():
06-25 00:41:56.038 28676 28676 E Python  :   File "/data/user/0/what.ever.acc_companion/files/python/user_code/app_packages/toga/widgets/textinput.py", line 172, in validate
06-25 00:41:56.038 28676 28676 E Python  :     for validator in self.validators:
06-25 00:41:56.038 28676 28676 E Python  :   File "/data/user/0/what.ever.acc_companion/files/python/user_code/app_packages/toga/widgets/textinput.py", line 132, in validators
06-25 00:41:56.038 28676 28676 E Python  :     return self._validators
06-25 00:41:56.038 28676 28676 E Python  : AttributeError: 'TextInput' object has no attribute '_validators'
```

The problem seems to be that the `on_change` callback method is called *before* the [validators are set](https://github.com/beeware/toga/blob/8b5f4392e16f45240da1dcd25120238a38fb92b8/src/core/toga/widgets/textinput.py#L51). I suspect that this `on_change` callback happens as part of [`self._create()`](https://github.com/beeware/toga/blob/8b5f4392e16f45240da1dcd25120238a38fb92b8/src/core/toga/widgets/textinput.py#L43) but I don't see why that would happen or why it is necessary (on Linux I don't observe this behavior); hence I suspect it's a bug.

**To Reproduce**

1. Create a text input field `field = toga.TextInput(on_change=my_callback, validators=['x'.__eq__])` where `my_callback` is defined as: `def my_callback(field): print(field.validate())`.
2. The crucial point here is that `field.validate()` is called inside the `on_change` callback (which seems to be invoked when the object is created).

**Expected behavior**

The `on_change` callback should not be called when the text input object is created.

**Environment:**
 - Operating System: Ubuntu 20.04 / Android 10
 - Python version: 3.9
 - Software versions:

```
arrow==1.1.0
binaryornot==0.4.4
briefcase==0.3.5
certifi==2021.5.30
chardet==4.0.0
click==8.0.1
cookiecutter==1.7.3
gbulb==0.6.1
gitdb==4.0.7
GitPython==3.1.17
idna==2.10
Jinja2==2.11.3
jinja2-time==0.2.0
MarkupSafe==2.0.1
mypy-extensions==0.4.3
poyo==0.5.0
pycairo==1.20.1
PyGObject==3.40.1
python-dateutil==2.8.1
python-slugify==5.0.2
requests==2.25.1
six==1.16.0
smmap==4.0.0
text-unidecode==1.3
toga-core==0.3.0.dev27
toga-gtk==0.3.0.dev27
toml==0.10.2
travertino==0.1.3
typed-ast==1.4.1
typing-extensions==3.7.4.2
urllib3==1.26.5
```

### LINKED ISSUE #1315: implementations should use interface values (vs raw)
**Describe the bug**
The code below explodes on iOS and Android (Mac/cocoa must do some casting?).

It works if the integers are cast via str().

```
        self.interval_input = toga.TextInput(
            placeholder=2, readonly=True, initial=2
        )
```

**To Reproduce**
1. Create an app with the code above
2. `briefcase run iOS` or `briefcase run android`
3. App will crash

**Expected behavior**
The code above should work.

**Screenshots**
n/a

**Environment:**
 - Operating System: Mac 11.4
 - Python version: 3.8
 - Software versions:
briefcase          0.3.5
toga               0.3.0.dev27
toga-cocoa         0.3.0.dev27
toga-core          0.3.0.dev27

### PR DESCRIPTION
**Pull request by freakboy3742**
Audit of both TextInput and PasswordInput (since they're essentially the same widget).

Extends #1938, as there's a lot of shared functionality in the probe.

* There are some backwards incompatible changes in the API of individual Validators. Since it's a fairly niche feature, that wasn't previously documented (although there was example code), I didn't think it was worth a full backwards compatibility path.

* Modifies the implementation of validation on macOS so that it doesn't require modal dialogs

* Adds confirmation handling on macOS

* Adds an implementation of focus handling, confirmation handling and validation on iOS.

Fixes #1290.
Fixes #1315.

## Audit checklist
 - [x] Core tests ported to pytest
 - [x] Core tests at 100% coverage 
 - [x] Core docstrings complete, and in Sphinx format
 - [x] Documentation complete and consistently formatted
 - [x] cocoa backend at 100% coverage 
 - [x] winforms backend at 100% coverage 
 - [x] gtk backend at 100% coverage 
 - [x] iOS backend at 100% coverage 
 - [x] android backend at 100% coverage 
 - [x] Widget support table updated
 - [x] Widget Issue backlog reviewed

### CODE REVIEW

**Comment by freakboy3742** [comment]
#1290 and #1315 should be fixed by this PR.

File: core/src/toga/validators.py [hunk 1]
```diff
@@ -147,43 +236,43 @@ def is_valid(self, input_string: str):
...
-        if isinstance(substrings, str):
-            substrings = [substrings]
-
+        """A validator confirming that the string contains one or more
+        substring.
+
+        :param substring: The substring that must exist in the input.
+        :param count: Optional; The exact number of matches that are expected.
+        :param error_message: Optional; the error message to display when the
+            input doesn't contain all the provided substrings.
```
Comment (mhsmith) [inline core/src/toga/validators.py:250]:
```suggestion
        :param error_message: Optional; the error message to display when the
            input doesn't contain the substring (or the requested count of substrings).
```

File: core/src/toga/validators.py [hunk 2]
```diff
@@ -220,151 +327,247 @@ def is_valid(self, input_string: str):
...
     ):
+        """A validator confirming that the string contains "special" (i.e.,
+        non-alphanumeric) characters.
+
+        :param count: Optional; if provided, the exact count of special
+            characters that must exist. If not provided, the existence of any
+            special character will make the string valid.
+        :param error_message: Optional; the error message to display when the
+            input doesn't contain special character (or the requested count of
+            special character).
```
Comment (mhsmith) [inline core/src/toga/validators.py:472]:
```suggestion
        :param error_message: Optional; the error message to display when the
            input doesn't contain special characters (or the requested count of
            special characters).
```

File: core/src/toga/validators.py [hunk 3]
```diff
@@ -13,98 +22,162 @@ def __call__(self, input_string: str):
...
+        self,
+        length: int,
+        error_message: Optional[str] = None,
+        allow_empty: bool = True,
     ):
+        """A validator confirming that the length of input does not exceed a maximum size.
+
+        :param length: The maximum length of the string (inclusive).
+        :param error_message: Optional; the error message to display when the
+            string isn't long enough.
```
Comment (mhsmith) [inline core/src/toga/validators.py:170]:
```suggestion
            string is too long.
```

File: winforms/src/toga_winforms/widgets/textinput.py [hunk 4]
```diff
@@ -120,6 +102,3 @@ def clear_error(self):
 
     def set_error(self, error_message):
         self.error_provider.SetError(self.native, error_message)
-
-    def winforms_double_click(self, sender, event):
-        self.native.SelectAll()
```
Comment (mhsmith) [inline winforms/src/toga_winforms/widgets/textinput.py:120]: I don't know why this was here: double-clicking on Windows is supposed to select one word, not the entire box.

File: winforms/src/toga_winforms/widgets/textinput.py [hunk 5]
```diff
@@ -74,37 +73,20 @@ def set_color(self, color):
...
-    def set_on_lose_focus(self, handler):
-        # No special handling required
-        pass
-
     def winforms_text_changed(self, sender, event):
         self.interface.on_change(self.interface)
         self.interface.validate()
 
-    def winforms_validated(self, sender, event):
-        self.interface.validate()
```
Comment (mhsmith) [inline winforms/src/toga_winforms/widgets/textinput.py]: I can't work out what the Validated event handler was supposed to do, but everything seems to work fine without it.

File: winforms/tests_backend/widgets/textinput.py [hunk 6]
```diff
@@ -1,7 +1,66 @@
...
+
+        SendKeys.SendWait(key_code)
+
+    @property
+    def alignment(self):
+        return toga_xalignment(self.native.TextAlign)
+
+    def assert_height(self, min_height, max_height):
+        # Height isn't configurable in this native widget.
+        assert 12 <= self.height <= 22
```
Comment (mhsmith) [inline winforms/tests_backend/widgets/textinput.py:62]: The native TextBox seems to ignore its height allocation and always stay the same height. Does this happen with any other widgets? I think I remember us talking about it, but I can't find any similar implementations of this method.

File: pyproject.toml [hunk 7]
```diff
@@ -48,3 +48,10 @@ filename = "docs/background/releases.rst"
 title_format = "{version} ({project_date})"
 issue_format = "`#{issue} <https://github.com/beeware/toga/issues/{issue}>`__"
 template = "changes/template.rst"
+type = [
+    { directory = "feature", name = "Features", showcontent = true },
+    { directory = "bugfix", name = "Bugfixes", showcontent = true },
+    { directory = "removal", name = "Backward Incompatible Changes", showcontent = true },
```
[the remark this answers is not in the document]
Comment (freakboy3742) [reply pyproject.toml:54]: I guess my take would be that you talk about the awesome new stuff before you scare people with the things that have changed :-)

As a comparable example - [Django's release notes](https://docs.djangoproject.com/en/4.2/releases/4.2/) list major features, minor features, then backwards incompatibilities.

File: core/src/toga/validators.py [hunk 8]
```diff
@@ -13,98 +22,162 @@ def __call__(self, input_string: str):
...
+        :param allow_empty: Optional; Is no input considered valid? Defaults to
+            ``True``
```
[the remark this answers is not in the document]
Comment (freakboy3742) [reply core/src/toga/validators.py:172]: True - I'll remove it.

File: core/src/toga/widgets/textinput.py [hunk 9]
```diff
@@ -1,212 +1,184 @@
...
         return self._on_lose_focus
 
     @on_lose_focus.setter
     def on_lose_focus(self, handler):
         self._on_lose_focus = wrapped_handler(self, handler)
-        self._impl.set_on_lose_focus(self._on_lose_focus)
 
-    def validate(self):
+    def validate(self) -> bool:
+        """Validate the current value of the widget.
```
[the remark this answers is not in the document]
Comment (freakboy3742) [reply core/src/toga/widgets/textinput.py:176]: No particular reason for it to be public (other than, at this point, backwards compatibility). At the very least, it's worth flagging in docs that end-users shouldn't ever need to call this method; but I wouldn't be opposed to making it an internal method (with the appropriate backwards compatibility note).

File: winforms/src/toga_winforms/widgets/textinput.py (hunk 4 above)
Comment (freakboy3742) [reply winforms/src/toga_winforms/widgets/textinput.py:120]: Yeah - I can't explain this one either.

File: winforms/tests_backend/widgets/textinput.py (hunk 6 above)
Comment (freakboy3742) [reply winforms/tests_backend/widgets/textinput.py:62]: There's some similar behavior with Button. Buttons don't expand to fill the available vertical space unless they're *explicitly* set to a given height. 

There's also some similar (but inconsistent) behavior with Switch. On most platforms, the switch "graphic" is a fixed height, regardless of the containing box; however, GTK and iOS both violate this, and render the switch with a weird aspect ratio.

In this case, I'm happy to call this a platform eccentricity and document it.

File: winforms/src/toga_winforms/widgets/textinput.py (hunk 5 above)
Comment (freakboy3742) [reply winforms/src/toga_winforms/widgets/textinput.py]: Winforms provides explicit event hooks to perform validation independent of focus and text change; I think the intention is that you can prevent a widget from losing focus if it's not currently valid. However, we're not blocking focus in that way, so I guess we can just piggyback on the change event.

**Review by mhsmith — APPROVED** [review]
@freakboy3742: All platforms now have 100% coverage, and a fully-working `textinput` example. Please review my last few commits, and merge the PR if you're happy.

**Comment by freakboy3742** [comment]
@mhsmith One minor tweak (avoiding the use of ObjC invocations for validation), but otherwise this all looks good to me.

**Pull request merged**